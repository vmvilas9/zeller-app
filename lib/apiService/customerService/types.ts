export type ZellerCustomer =  {
    id: string;
    name: string;
    email: string;
    role: string;
}

export type ZellerCustomerConnection = {
    items: ZellerCustomer[];
    nextToken: string;
}

export type TableStringFilterInput = {
    ne?: string;
    eq?: string;
    le?: string;
    lt?: string;
    ge?: string;
    gt?: string;
    contains?: string;
    notContains?: string;
    between?: [string, string];
    beginsWith?: string;
};

export type TableZellerCustomerFilterInput = {
    id?: TableStringFilterInput;
    name?: TableStringFilterInput;
    email?: TableStringFilterInput;
    role?: TableStringFilterInput;
};
