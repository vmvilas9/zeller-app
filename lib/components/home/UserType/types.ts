export type UserType = {
    displayName: string;
    type: string;
}
