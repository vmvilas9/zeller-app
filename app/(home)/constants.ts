import {UserType} from "@/lib/components/home/UserType/types";

export const userTypes: UserType[] = [
    {displayName: "Admin", type: "ADMIN"},
    {displayName: "Manager", type: "MANAGER"}
];
