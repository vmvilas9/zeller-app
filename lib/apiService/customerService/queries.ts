import { gql } from "@apollo/client";

export const GET_ZELLER_CUSTOMER = gql`
  query GetZellerCustomer($id: String!) {
    getZellerCustomer(id: $id) {
      id
      name
      email
      role
    }
  }
`;

export const LIST_ZELLER_CUSTOMERS = gql`
  query ListZellerCustomers(
    $filter: TableZellerCustomerFilterInput
    $limit: Int
    $nextToken: String
  ) {
    listZellerCustomers(filter: $filter, limit: $limit, nextToken: $nextToken) {
      items {
        id
        name
        email
        role
      }
      nextToken
    }
  }
`;
