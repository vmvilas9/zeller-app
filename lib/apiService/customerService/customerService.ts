import {ApolloQueryResult} from "@apollo/client";
import client from "../../network/client";
import {GET_ZELLER_CUSTOMER, LIST_ZELLER_CUSTOMERS} from "./queries";
import {TableZellerCustomerFilterInput, ZellerCustomer, ZellerCustomerConnection} from "./types";

export async function fetchZellerCustomer(id: string): Promise<ZellerCustomer | undefined> {
    try {
        const result: ApolloQueryResult<{ getZellerCustomer: ZellerCustomer }> = await client.query({
            query: GET_ZELLER_CUSTOMER,
            variables: {id},
        });
        return result.data?.getZellerCustomer;
    } catch (error) {
        console.error("Error fetching customer:", error);
        return undefined;
    }
}

export async function listZellerCustomers (
    filter?: TableZellerCustomerFilterInput,
    limit: number = 50,
    nextToken?: string
): Promise<ZellerCustomerConnection> {
    try {
        const result: ApolloQueryResult<{
            listZellerCustomers: ZellerCustomerConnection;
        }> = await client.query({
            query: LIST_ZELLER_CUSTOMERS,
            variables: {filter, limit, nextToken},
        });

        return result.data?.listZellerCustomers || {items: [], nextToken: null};
    } catch (error) {
        console.error("Error listing customers:", error);
        throw error;
    }
}
