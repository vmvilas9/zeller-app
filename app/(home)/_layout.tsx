import React, {useEffect, useState} from 'react';
import {FlatList} from "react-native";
import {listZellerCustomers} from "@/lib/apiService/customerService/customerService";
import {TableZellerCustomerFilterInput, ZellerCustomer} from "@/lib/apiService/customerService/types";
import {userTypes} from "@/app/(home)/constants";
import {useRouter} from "expo-router";
import {MyActivityIndicator, MySafeAreaView, MyText, MyView, UserTypeView, User, Divider} from "@/lib/components";

export default function Home() {
    const router = useRouter()
    const [selectedUserType, setSelectedUserType] = useState<string>(userTypes[0].type);
    const [loading, setLoading] = useState(false);
    const [customers, setCustomers] = useState<ZellerCustomer[]>([]);

    useEffect(() => {
        const filter: TableZellerCustomerFilterInput = {
            role: {eq: selectedUserType},
        };
        const fetchCustomers = async () => {
            setLoading(true);
            try {
                const data = await listZellerCustomers(filter);
                setCustomers(data.items);
            } catch (error) {
                console.error(error);
            } finally {
                setLoading(false);
            }
        };
        fetchCustomers();
    }, [selectedUserType]);

    const navigateToUserDetailsPage = (userId: string) => {
        router.push(`/(userDetails)/${userId}`);
    }

    return (
        <MySafeAreaView>
            <MyView>
                <MyText type={'defaultSemiBold'} style={{marginBottom: 10}}>User Types</MyText>
                {userTypes.map((type) => (
                    <UserTypeView
                        key={type.type}
                        userType={type}
                        selectedUserType={selectedUserType}
                        setSelectedUserType={setSelectedUserType}/>))}
                <Divider/>
                <MyText type={'defaultSemiBold'} style={{marginBottom: 10}}>{selectedUserType} Users</MyText>
                {loading ?
                    <MyActivityIndicator testID="activity-indicator"/> :
                    <FlatList
                        testID={"user-types"}
                        data={customers}
                        keyExtractor={(item) => item.id.toString()}
                        renderItem={({item}) => (<User userDetails={item} onUserClick={navigateToUserDetailsPage}/>)}
                        pagingEnabled
                    />}
            </MyView>
        </MySafeAreaView>
    );
};
