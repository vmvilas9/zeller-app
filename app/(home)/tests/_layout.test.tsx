import {render, fireEvent, waitFor} from '@testing-library/react-native';
import Home from '../_layout';  // Adjust the path as necessary
import {listZellerCustomers} from '@/lib/apiService/customerService/customerService';  // Adjust the path as necessary
import {userTypes} from "@/app/(home)/constants";
import {TableZellerCustomerFilterInput} from "@/lib/apiService/customerService/types";
import React from "react";
import {useRouter} from "expo-router";  // Adjust the path as necessary

// Mock API call
jest.mock('@/lib/apiService/customerService/customerService', () => ({
    listZellerCustomers: jest.fn(),
}));

// Mock useRouter hook for navigation
jest.mock('expo-router', () => ({
    useRouter: jest.fn(),
}));

describe('Home Component', () => {
    const filter: TableZellerCustomerFilterInput = {
        role: {eq: 'ADMIN'},
    };
    const mockCustomers = [{id: '1', name: 'John Doe', role: 'ADMIN', email: 'john@example.com'}];

    it('renders the Home component correctly', async () => {
        (listZellerCustomers as jest.Mock).mockReturnValue(filter);
        (listZellerCustomers as jest.Mock).mockResolvedValueOnce({items: mockCustomers});
        const {toJSON, queryByTestId} = render(<Home/>);
        await waitFor(() => expect(queryByTestId('activity-indicator')).toBeNull());
        expect(toJSON()).toMatchSnapshot();
    });

    it('renders loading indicator when data is being fetched', async () => {
        (listZellerCustomers as jest.Mock).mockReturnValue(filter);
        (listZellerCustomers as jest.Mock).mockResolvedValueOnce({items: mockCustomers});
        const {findByTestId} = render(<Home/>);

        // Simulate loading state
        expect(await findByTestId('activity-indicator')).toBeTruthy();
    });

    it('renders customer list after data is fetched', async () => {
        (listZellerCustomers as jest.Mock).mockReturnValue(filter);
        (listZellerCustomers as jest.Mock).mockResolvedValueOnce({items: mockCustomers});

        const {findByTestId, findByText} = render(<Home/>);

        // Wait for the list to be rendered
        await waitFor(() => expect(findByTestId('user-types')).toBeTruthy());

        // Check if the customer name is rendered
        expect(findByText('John Doe')).toBeTruthy();
    });

    it('fetches customers based on selected user type', async () => {
        (listZellerCustomers as jest.Mock).mockReturnValue(filter);
        (listZellerCustomers as jest.Mock).mockResolvedValueOnce({items: mockCustomers});

        const {getByText, queryByTestId, findByText} = render(<Home/>);

        // Select a different user type
        fireEvent.press(getByText('Admin'));

        await waitFor(() => expect(queryByTestId('activity-indicator')).toBeNull());

        await waitFor(() => expect(queryByTestId('user-types')).toBeTruthy());

        // Wait for the customer data to be rendered after filtering
        await waitFor(() => expect(findByText('John Doe')).toBeTruthy());
    });

    it('navigates to user details page when a user is clicked', async () => {
        const mockPush = jest.fn();
        (listZellerCustomers as jest.Mock).mockReturnValue(filter);
        (useRouter as jest.Mock).mockReturnValue({push: mockPush});


        (listZellerCustomers as jest.Mock).mockResolvedValueOnce({items: mockCustomers});

        const {getByTestId, queryByTestId} = render(<Home/>);

        await waitFor(() => expect(queryByTestId('user-types')).toBeTruthy());

        // Trigger the click event
        fireEvent.press(getByTestId('user-item'));

        // Check if the navigation function is called with the correct user ID
        expect(mockPush).toHaveBeenCalledWith('/(userDetails)/1');
    });

    it('renders user types and updates customer list when a type is selected', () => {
        const {findByText, getByText} = render(<Home/>);

        // Check if user types are rendered
        userTypes.forEach((type) => {
            expect(findByText(type.displayName)).toBeTruthy();
        });

        // Simulate selecting a different user type
        fireEvent.press(getByText(userTypes[1].displayName));  // Selecting second user type

        // Check if the customer list is filtered accordingly
        expect(findByText(`${userTypes[1].type} Users`)).toBeTruthy();
    });

});
