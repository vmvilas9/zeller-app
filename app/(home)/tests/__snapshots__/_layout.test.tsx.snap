// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component renders the Home component correctly 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "additive",
      "left": "additive",
      "right": "additive",
      "top": "additive",
    }
  }
  style={
    {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "padding": 30,
    }
  }
>
  <View
    style={
      [
        {
          "backgroundColor": "#FFFFFF",
        },
        undefined,
      ]
    }
  >
    <Text
      style={
        [
          undefined,
          undefined,
          {
            "fontSize": 16,
            "fontWeight": "600",
            "lineHeight": 24,
          },
          undefined,
          undefined,
          {
            "marginBottom": 10,
          },
        ]
      }
    >
      User Types
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#E8F2FB",
          "borderRadius": 8,
          "flexDirection": "row",
          "marginTop": 10,
          "opacity": 1,
          "paddingHorizontal": 16,
          "paddingVertical": 8,
        }
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "#FFFFFF",
            },
            {
              "alignItems": "center",
              "borderColor": "#0171CE",
              "borderRadius": 12,
              "borderWidth": 2,
              "height": 24,
              "justifyContent": "center",
              "marginRight": 12,
              "width": 24,
            },
          ]
        }
        testID="radio-outer-circle"
      >
        <View
          style={
            [
              {
                "backgroundColor": "#FFFFFF",
              },
              {
                "backgroundColor": "#0171CE",
                "borderRadius": 6,
                "height": 12,
                "width": 12,
              },
            ]
          }
          testID="radio-inner-circle"
        />
      </View>
      <Text
        style={
          [
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
          ]
        }
      >
        Admin
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "borderRadius": 8,
          "flexDirection": "row",
          "marginTop": 10,
          "opacity": 1,
          "paddingHorizontal": 16,
          "paddingVertical": 8,
        }
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "#FFFFFF",
            },
            {
              "alignItems": "center",
              "borderColor": "#0171CE",
              "borderRadius": 12,
              "borderWidth": 2,
              "height": 24,
              "justifyContent": "center",
              "marginRight": 12,
              "width": 24,
            },
          ]
        }
        testID="radio-outer-circle"
      />
      <Text
        style={
          [
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
          ]
        }
      >
        Manager
      </Text>
    </View>
    <View
      style={
        [
          {
            "backgroundColor": "#FFFFFF",
          },
          {
            "backgroundColor": "#E2E8F0",
            "height": 1,
            "marginVertical": 24,
          },
        ]
      }
    />
    <Text
      style={
        [
          undefined,
          undefined,
          {
            "fontSize": 16,
            "fontWeight": "600",
            "lineHeight": 24,
          },
          undefined,
          undefined,
          {
            "marginBottom": 10,
          },
        ]
      }
    >
      ADMIN
       Users
    </Text>
    <RCTScrollView
      data={
        [
          {
            "email": "john@example.com",
            "id": "1",
            "name": "John Doe",
            "role": "ADMIN",
          },
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      pagingEnabled={true}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={0.0001}
      stickyHeaderIndices={[]}
      testID="user-types"
      viewabilityConfigCallbackPairs={[]}
    >
      <View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
            testID="user-item"
          >
            <View
              style={
                [
                  {
                    "backgroundColor": "#FFFFFF",
                  },
                  {
                    "alignItems": "center",
                    "borderBottomColor": "#EEEEEE",
                    "borderBottomWidth": 1,
                    "flexDirection": "row",
                    "paddingVertical": 12,
                  },
                ]
              }
            >
              <View
                style={
                  [
                    {
                      "backgroundColor": "#FFFFFF",
                    },
                    {
                      "alignItems": "center",
                      "backgroundColor": "#E8F2FB",
                      "borderRadius": 4,
                      "height": 40,
                      "justifyContent": "center",
                      "marginRight": 12,
                      "width": 40,
                    },
                  ]
                }
              >
                <Text
                  style={
                    [
                      {
                        "fontSize": 16,
                        "lineHeight": 24,
                      },
                      undefined,
                      undefined,
                      undefined,
                      undefined,
                      {
                        "color": "#0171CE",
                      },
                    ]
                  }
                  testID="avatar-text"
                >
                  J
                </Text>
              </View>
              <View
                style={
                  [
                    {
                      "backgroundColor": "#FFFFFF",
                    },
                    {
                      "flex": 1,
                    },
                  ]
                }
              >
                <Text
                  style={
                    [
                      {
                        "fontSize": 16,
                        "lineHeight": 24,
                      },
                      undefined,
                      undefined,
                      undefined,
                      undefined,
                      undefined,
                    ]
                  }
                >
                  John Doe
                </Text>
                <Text
                  style={
                    [
                      {
                        "fontSize": 16,
                        "lineHeight": 24,
                      },
                      undefined,
                      undefined,
                      undefined,
                      undefined,
                      {
                        "color": "#777777",
                      },
                    ]
                  }
                >
                  ADMIN
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</RNCSafeAreaView>
`;
